### PR TITLE
SLCORE-89 Allow to connect to SonarQube server with incompatible plugin

### PR DESCRIPTION
**Pull request by dbmeneses**

### CODE REVIEW

File: core/src/main/java/org/sonarsource/sonarlint/core/container/storage/StorageContainer.java [hunk 1]
```diff
@@ -138,6 +142,17 @@ public GlobalStorageStatus getGlobalStorageStatus() {
...
 
+  public List<LoadedAnalyzer> getAnalyzers() {
+    DefaultPluginRepository pluginRepository = getComponentByType(DefaultPluginRepository.class);
+    return pluginRepository.getPluginInfos().stream()
+      .map(StorageContainer::pluginInfoToAnalyzer)
+      .collect(Collectors.toList());
+  }
+
+  private static LoadedAnalyzer pluginInfoToAnalyzer(PluginInfo p) {
+    return new DefaultLoadedAnalyzer(p.getKey(), p.getName(), p.getVersion() == null ? null : p.getVersion().toString());
```
Comment (SonarTech) [inline core/src/main/java/org/sonarsource/sonarlint/core/container/storage/StorageContainer.java]: MAJOR A "NullPointerException" could be thrown; "getVersion" is nullable here [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2259)

File: core/src/main/java/org/sonarsource/sonarlint/core/container/connected/update/PluginListDownloader.java [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+import org.sonarsource.sonarlint.core.container.connected.SonarLintWsClient;
+import org.sonarsource.sonarlint.core.container.connected.validate.PluginVersionChecker;
+import org.sonarsource.sonarlint.core.container.model.DefaultSonarAnalyzer;
+import org.sonarsource.sonarlint.core.plugin.PluginCopier;
+import org.sonarsource.sonarlint.core.plugin.Version;
+import org.sonarsource.sonarlint.core.util.VersionUtils;
+import org.sonarsource.sonarlint.core.util.ws.WsResponse;
+
+public class PluginListDownloader {
+  public static final String WS_PATH = "/api/plugins/installed";
```
Comment (SonarTech) [inline core/src/main/java/org/sonarsource/sonarlint/core/container/connected/update/PluginListDownloader.java]: MINOR Refactor your code to get this URI from a customizable parameter. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1075)

File: core/src/main/java/org/sonarsource/sonarlint/core/container/connected/update/PluginListDownloader.java [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+  public static final String WS_PATH_LTS = "/deploy/plugins/index.txt";
```
Comment (SonarTech) [inline core/src/main/java/org/sonarsource/sonarlint/core/container/connected/update/PluginListDownloader.java]: MINOR Refactor your code to get this URI from a customizable parameter. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1075)

File: core/src/main/java/org/sonarsource/sonarlint/core/container/connected/update/PluginReferencesDownloader.java [hunk 4]
```diff
@@ -19,75 +19,71 @@
...
+          .build());
+      }
     }
-    scanner.close();
     return builder.build();
   }
 
-  public PluginReferences fetchPluginsTo(Path dest, String serverVersion) {
-    PluginReferences refs = fetchPlugins(serverVersion);
+  private boolean checkVersion(@Nullable String version, @Nullable String minVersion) {
```
Comment (SonarTech) [inline core/src/main/java/org/sonarsource/sonarlint/core/container/connected/update/PluginReferencesDownloader.java]: MAJOR Make "checkVersion" a "static" method. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2325)

File: core/src/main/java/org/sonarsource/sonarlint/core/container/connected/update/PluginReferencesDownloader.java [hunk 5]
```diff
@@ -19,75 +19,71 @@
...
-      String[] nameAndHash = StringUtils.split(fields[fields.length - 1], "|");
-      String key = fields[0];
-      boolean compatible = PluginCopier.isWhitelisted(key) || !compatibleFlagPresent || "true".equals(fields[1]);
-      if (!compatible) {
-        LOG.debug("Plugin {} is not compatible with SonarLint. Skip it.", key);
+  public PluginReferences fetchPlugins(List<SonarAnalyzer> analyzers) {
+    Builder builder = PluginReferences.newBuilder();
+    for (SonarAnalyzer analyzer : analyzers) {
+      if (!analyzer.sonarlintCompatible()) {
+        LOG.debug("Plugin {} is not compatible with SonarLint. Skip it.", analyzer.key());
```
Comment (SonarTech) [inline core/src/main/java/org/sonarsource/sonarlint/core/container/connected/update/PluginReferencesDownloader.java]: MAJOR Invoke method(s) only conditionally. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2629)

**Comment by SonarTech** [comment]
SonarQube analysis reported 5 issues
* ![MAJOR][MAJOR] 3 major
* ![MINOR][MINOR] 2 minor

Watch the comments in this conversation to review them.

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: core/src/main/java/org/sonarsource/sonarlint/core/ConnectedSonarLintEngineImpl.java [hunk 6]
```diff
@@ -192,12 +197,19 @@ public GlobalStorageStatus update(ServerConfiguration serverConfig, @Nullable Pr
         }
         start();
       }
-      return getGlobalContainer().getGlobalStorageStatus();
+      return new UpdateResult(getGlobalContainer().getGlobalStorageStatus(), analyzers);
```
Comment (janos-ss) [inline core/src/main/java/org/sonarsource/sonarlint/core/ConnectedSonarLintEngineImpl.java]: It looks like `analyzers` might not be initialized at this point.

File: core/src/main/java/org/sonarsource/sonarlint/core/plugin/PluginInfo.java [hunk 7]
```diff
@@ -36,7 +36,7 @@
 import org.sonar.api.utils.MessageException;
 import org.sonar.api.utils.log.Loggers;
 
-public class PluginInfo implements Comparable<PluginInfo> {
+public class PluginInfo implements Comparable<PluginInfo>{
```
Comment (janos-ss) [inline core/src/main/java/org/sonarsource/sonarlint/core/plugin/PluginInfo.java]: Junk change? ;-)

**Comment by SonarTech** [comment]
SonarQube analysis reported 4 issues
* ![MAJOR][MAJOR] 2 major
* ![MINOR][MINOR] 2 minor

Watch the comments in this conversation to review them.

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Pull request merged**